### ISSUE TITLE
send_arp.linux ignores arguments

### ISSUE THREAD

**Issue by phemmer** [ClusterLabs/resource-agents#860]
The [`send_arp` utility for linux](https://github.com/ClusterLabs/resource-agents/blob/master/tools/send_arp.linux.c) ignores the `src_hw_addr`, `broadcast_ip_addr`, and `netmask` arguments. This results in the utility sending out the wrong mac address when called by `IProute2` in clone (clusterip) mode.

https://github.com/ClusterLabs/resource-agents/blob/master/tools/send_arp.linux.c#L1105-L1115
^ note that the only parameters that are actually used are `device` and `src_ip_addr`.

Also, on an unrelated note, the usage documentation is completely wrong: https://github.com/ClusterLabs/resource-agents/blob/master/tools/send_arp.linux.c#L140-L150

**Comment by lge** [comment]
> The [`send_arp` utility for
> linux](https://github.com/ClusterLabs/resource-agents/blob/master/tools/send_arp.linux.c)
> ignores the `src_hw_addr`, `broadcast_ip_addr`, and `netmask`
> arguments. This results in the utility sending out the wrong mac
> address when called by `IProute2` in clone (clusterip) mode.

Depends on exact configuration, but generally: yes, you are right.

> https://github.com/ClusterLabs/resource-agents/blob/master/tools/send_arp.linux.c#L1105-L1115
> ^ note that the only parameters that are actually used are `device` and `src_ip_addr`.

"Always been broken that way" :-(

> Also, on an unrelated note, the usage documentation is completely
> wrong:
> https://github.com/ClusterLabs/resource-agents/blob/master/tools/send_arp.linux.c#L140-L150

For the interested, let me add some historic context:
on "inception", heartbeat came with its own, homegrown, libnet based
send_arp tool (because apparently a working, cross-platform, reliably
available tools with similar functionality.

Its ancestor is still available as the "send_arp.libnet.c" variant.
Later, the libnet dependency was found to be undesirable (for whatever
reason), so on linux at least, the iputils arping was canibalized,
and put into send_arp.linux.c, with mangling of the options parsing,
and added "pidfile", so it could be used as a drop-in replacement of its
existing send_arp.libnet tool.

The usage() displayed is that of iputils arping,
but the way it is used by IPaddr2 is how send_arp.libnet.c was used.
It basically has two usages, but the project uses the one not documented.

Last time I touched IPaddr2, I suggested that we would instead
change the send_arp.libnet.c variant to follow the arping command line
instead, and phase out our "in-project" send_arp utility, rather relying
on a platform installed usable arping instead (where available).

But no-one did actually do that.

Anyways, apparently the "use an explicit caller specified MAC" (instead
of the auto-magically guessed from NIC one) was patched into send_arp.libnet.c 
but never into send_arp.linux.c.

==================

I suggest that we finally really change the command line to the arping one, 
change send_arp.libnet.c for those platforms that still need it to
understand it that way.

And not ignore command line parameters ;-)



-- 
: Lars Ellenberg
: LINBIT | Keeping the Digital World Running
: DRBD -- Heartbeat -- Corosync -- Pacemaker
: R&D, Integration, Ops, Consulting, Support

DRBD� and LINBIT� are registered trademarks of LINBIT
